src/code/sys_flashrom.c

```diff
@@ -3,42 +3,242 @@
+    if (result != 0) {
+        if (i < 3) {
+            i++;
+            goto failRetry;
+        }
+        return result;
+    }
+    return 0;
+}
+
+s32 func_80185BE4(void* data, void* addr, u32 pageCount) {
+    u32 size;
+    u32 i;
+
+    for (i = 0; i < pageCount * 128; i += 4) {
+        if ((*(s32*)data & *(s32*)addr) != *(s32*)addr) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+s32 SysFlashrom_WriteData2(void* addr, u32 pageNum, u32 pageCount) {
+    void* data;
+    size_t size;
+    s32 ret;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    size = pageCount * 128;
+    data = SystemArena_Malloc(size);
+    if (data == 0) {
+        ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+    } else {
+        SysFlashrom_ReadData(data, pageNum, pageCount);
+        if (bcmp(data, addr, size) == 0) {
+            ret = 0;
+        } else {
+            //! @bug Will attempt to write regardless of the result of the function call.
+            if (func_80185BE4(data, addr, pageCount) != 0) {
+                ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+            } else {
+                ret = SysFlashrom_AttemptWrite(addr, pageNum, pageCount);
+            }
+            if (ret == 0) {
+                SysFlashrom_ReadData(data, pageNum, pageCount);
+                if (bcmp(data, addr, size) == 0) {
+                    ret = 0;
+                } else {
+                    ret = -1;
+                }
+            }
+        }
+        SystemArena_Free(data);
+    }
+    return ret;
+}
+
+void SysFlashrom_ThreadEntry(void* arg) {
+    FlashromRequest* req = (FlashromRequest*)arg;
+    switch (req->requestType) {
+        case 1:
+            req->response = SysFlashrom_WriteData2(req->addr, req->pageNum, req->pageCount);
+            osSendMesg(&req->messageQueue, req->response, OS_MESG_BLOCK);
+            break;
+        case 2:
+            req->response = SysFlashrom_ReadData(req->addr, req->pageNum, req->pageCount);
+            osSendMesg(&req->messageQueue, req->response, OS_MESG_BLOCK);
+            break;
+    }
+}
+
+void SysFlashrom_CreateRequest(u8* addr, u32 pageNum, u32 pageCount) {
+    FlashromRequest* req = &sFlashromRequest;
+    if (SysFlashrom_FlashIsGood() != 0) {
+        req->requestType = 1;
+        req->addr = addr;
+        req->pageNum = pageNum;
+        req->pageCount = pageCount;
+        osCreateMesgQueue(&req->messageQueue, sSysFlashromMsgBuf, ARRAY_COUNT(sSysFlashromMsgBuf));
+        StackCheck_Init(&sSysFlashromStackInfo, sSysFlashromStack, STACK_TOP(sSysFlashromStack), 0,
+                        0x100, "sys_flashrom");
+        osCreateThread(&sSysFlashromThread, Z_THREAD_ID_FLASHROM, SysFlashrom_ThreadEntry, req,
+                       STACK_TOP(sSysFlashromStack), Z_PRIORITY_FLASHROM);
+        osStartThread(&sSysFlashromThread);
+    }
+}
+
+s32 SysFlashrom_IsQueueFull(void) {
```

Comment:
```suggestion
s32 SysFlashrom_IsBusy(void) {
```
The queue is one message large so more generally this is just a check for whether the flash thread is currently working or not